WIP: Refactor to move gating and extra-attrs logic to PostsService

kevinansfield opened the pull request:
ref https://linear.app/ghost/issue/PLG-352

- updated `/p/` endpoint to accept an optional `?memberStatus` query parameter with three allowed values: `anonymous/free/paid`
  - using a query parameter allows for post previews to be opened in a new tab and still display segmented views
- when a `memberStatus` query param is provided:
  - `frame.apiType` is set to `content` which triggers the content-gating paths of our output serializer's post mapper
  - where needed, `frame.original.context.member` is set to a minimal member object with a `status` property to allow content-gating NQL queries to match
  - the combination of both switches our preview API output to match what the Content API does when rendering posts with a particular logged-in member status
- if no `memberStatus` query param is provided the existing behaviour is preserved, where we render as we do for Admin API requests, showing content for all segments simultaneously

kevinansfield committed:
Added preview-as anonymous/free/paid to previews endpoint

ref https://linear.app/ghost/issue/PLG-352

- updated `/p/` endpoint to accept an optional `?memberStatus` query parameter with three allowed values: `anonymous/free/paid`
- when a `memberStatus` query param is provided:
  - `frame.apiType` is set to `content` which triggers the content-gating paths of our output serializer's post mapper
  - where needed, `frame.original.content.member` is set to a minimal member object with a `status` property to allow content-gating NQL queries to match
  - the combination of both changes our API output to match what the Content API does when rendering posts with a particular logged-in member status
- if no `memberStatus` query param is provided the existing behaviour is preserved of rendering as we do for Admin API requests, showing content for all segments simultaneously

kevinansfield committed:
Extracted content gating checks to content-gating service

no issue

- content gating lived on the members service despite not having any direct ties to it, as a result we also had content gating related logic split across the members service and post output serializer
- extracted existing logic from the members service to a new `content-gating` service and updated post output serializer "post-gating" to use the new code location
- first step towards a more comprehensive refactor

kevinansfield committed:
Extracted gated blocks parsing and manipulation to content-gating service

no issue

- logic for handling of gated blocks lived in the post output serializer's "post gating" module which meant it was disjointed from other content gating related code and was difficult to re-use in other areas
- moved the logic to the new `content-gating` service to keep related functionality co-located and reusable

kevinansfield committed:
Extracted post attr gating to content-gating service

no issue

- moved remaining gating logic to content-gating service
- kept modify-by-reference to avoid deep cloning post objects for performance on a hot path
- cleaned up requires to avoid circular dependencies and function wrappers

ronaldlangeveld commented on ghost/core/core/server/api/endpoints/previews.js:
```diff
@@ -2,11 +2,39 @@ const tpl = require('@tryghost/tpl');
+    if (!frame?.options?.memberStatus) {
+        return;
+    }
+
```
Would it make things safer if we add member validation here, eg checking that `frame.options.memberStatus` has one of `ALLOWED_MEMBER_STATUSES` and perhaps throw an error or default? In case

kevinansfield replied:
Yeah, we could potentially add some extra safety here. Although in theory it should never be possible as the [API filters the memberStatus option](https://github.com/TryGhost/Ghost/pull/22168/files#diff-f5713173a4138c9fb6f492e66ff27cad61e8250ecce7f13e7b7e33f39cec3411R59-R60) before this code is reached.

kevinansfield committed:
Updated API controllers to consistently use postsService

no issue

- changed `PostsService.browse` to match `PostsService.read` arguments, accepting a full page object
- previously `PostsService.browse` returned raw models from `findPage` whereas `PostsService.read` returned an already `.toJSON()`'d POJO
  - updated browse to match read behaviour which will let us gate attributes in the service rather than in the output serializer
  - modified posts output serializer & mapper to properly handle POJO's in place of models
  - an upshot of this is that our API will now consistently return an `id` property for posts even when not requested via `?fields`, previously the behaviour was inconsistent depending on whether the endpoint used the posts service
- updated posts-public.read to follow same pattern as browse endpoint using the postsService
- updated pages, pages-public, and email-posts controllers to use the postsService
  - allowed for removal of duplicate error message handling as it already exists on the posts service

kevinansfield committed:
Extracted post gating from output serializer to posts-service

no issue

- moves final piece of content gating logic out of the output serializer and into the posts-service to make things more explicit
- updates `previews` endpoint to use `posts-service` rather than direct model access because there's no longer automatic content gating at the API output serializer layer

kevinansfield committed:
WIP: move extra-attrs behaviour from mapper to posts-service

no issue

- the extra-attrs behaviour relies on receiving a model in order to access attributes that may not be included in `.toJSON` but that doesn't work if the mapper is passed a POJO such as when posts-service is used by controllers
- moving the attr modifying behaviour to `posts-service` keeps post attr logic centralized and allows us to ensure the post attrs are complete before passing through content-gating

TODO:
- [ ] finish moving functionality
- [ ] add `createPost` to `PostsService` and update controllers to use it
  - with the `extra-attrs` behaviour moved from the mapper to the service, we need to ensure we always go through the service when the data will be passed over for API output
- [ ] update `PostsService.editPost` to use the service's internal serialization and ensure controllers use it

ErisDS commented:
Cleaning up older PRs 💅
